src/chopped_graph.hpp

```diff
@@ -0,0 +1,850 @@
+    using ChoppedGraph<MaxNodeLength>::chopped_size;
+    using ChoppedGraph<MaxNodeLength>::chopped_index_bit_count;
+    using ChoppedGraph<MaxNodeLength>::chopped_index_mask;
+    using ChoppedGraph<MaxNodeLength>::chopped_handle;
+};
+
+/**
+ * The same as ChoppedPathGraph except with the PathPositionHandleGraph interface added
+ */
+template<uint64_t MaxNodeLength = 1024>
+class ChoppedPathPositionGraph : public ChoppedPathGraph<MaxNodeLength>, virtual public PathPositionHandleGraph {
+public:
+    
+    ChoppedPathPositionGraph(const PathPositionHandleGraph& graph);
+    ChoppedPathPositionGraph() = default;
+    virtual ~ChoppedPathPositionGraph() = default;
+    
+    /// Returns the length of a path measured in bases of sequence.
+    size_t get_path_length(const path_handle_t& path_handle) const;
+    
+    /// Returns the position along the path of the beginning of this step measured in
+    /// bases of sequence. In a circular path, positions start at the step returned by
+    /// path_begin().
+    size_t get_position_of_step(const step_handle_t& step) const;
+    
+    /// Returns the step at this position, measured in bases of sequence starting at
+    /// the step returned by path_begin(). If the position is past the end of the
+    /// path, returns path_end().
+    step_handle_t get_step_at_position(const path_handle_t& path, const size_t& position) const;
+    
+protected:
+    
+    const PathPositionHandleGraph* path_position_graph;
+    
+    using ChoppedPathGraph<MaxNodeLength>::chopped_step_handle;
+    using ChoppedPathGraph<MaxNodeLength>::chop_index;
+    using ChoppedPathGraph<MaxNodeLength>::get_underlying_step_handle;
+};
+
+
+
+
+
+
+
+
+/**
+ *
+ * Template implementations
+ *
+ */
+
+template<uint64_t MaxNodeLength>
+ChoppedGraph<MaxNodeLength>::ChoppedGraph(const HandleGraph& graph) : graph(&graph) {
+    
+    // note: assumes dense IDs starting at 1 for efficiency
+    // make a vector where every ID can also index before and after by 1
+    chopped_id_begin.resize(graph.max_node_id() + 2, 0);
+    size_t max_chopped_count = 0;
+    graph.for_each_handle([&](const handle_t& handle) {
+        // assign it an interval of IDs for the chopped nodes
+        size_t chopped_count = max<size_t>(1, (graph.get_length(handle) + MaxNodeLength - 1) / MaxNodeLength);
+        // temporarily assign only the count in the vector
+        chopped_id_begin[graph.get_id(handle) + 1] = chopped_count;
+        max_chopped_count = max(chopped_count, max_chopped_count);
+    });
+    
+    // convert it into a prefix sum
+    chopped_id_begin[0] = 1; // make IDs start at 1
+    for (size_t i = 1; i < chopped_id_begin.size(); ++i) {
+        chopped_id_begin[i] += chopped_id_begin[i - 1];
+    }
+    
+    // inefficient way to find most significant bit, but whatever
+    for (uint8_t i = 1; i <= 64; ++i) {
+        if (max_chopped_count & (1 << (i - 1))) {
+            chopped_index_bit_count = i;
```

Comment:
Since we have all the backing graph handles around, should we also check to make sure the handles of the backing graph leave enough high bits for the shift to be safe? We're starting to steal more than just one low bit now.

Reply:
Yeah, that's a good idea.